core/src/test/java/io/parsingdata/metal/AutoEqualityTest.java

```diff
@@ -0,0 +1,283 @@
+            Len.class, Offset.class, Neg.class, Not.class, Count.class, First.class, Last.class, Reverse.class,
+            And.class, Or.class, ShiftLeft.class, ShiftRight.class, Add.class, Div.class, Mod.class, Mul.class,
+            io.parsingdata.metal.expression.value.arithmetic.Sub.class, Cat.class, Eq.class, EqNum.class, EqStr.class,
+            GtNum.class, LtNum.class, Nth.class, Elvis.class, io.parsingdata.metal.expression.logical.And.class,
+            io.parsingdata.metal.expression.logical.Or.class, FoldLeft.class, FoldRight.class, Value.class,
+            ParseValue.class, io.parsingdata.metal.expression.logical.Not.class, Const.class, NameRef.class,
+            io.parsingdata.metal.expression.value.reference.TokenRef.class, ConstantSource.class, Slice.class,
+            ParseReference.class, DataExpressionSource.class, ParseGraph.class, ByteStreamSource.class
+        );
+    }
+
+    private static Collection<Object[]> generateObjectArrays(Class... classes) {
+        Collection<Object[]> results = new ArrayList<>();
+        for (Class c : classes) {
+            results.add(generateObjectArrays(c));
+        }
+        return results;
+    }
+
+    private static Object[] generateObjectArrays(Class c) {
+        Constructor cons = c.getConstructors()[0];
+        List<List<Object>> args = new ArrayList<>();
+        for (Class cl : cons.getParameterTypes()) {
+            args.add(mapping.get(cl));
+        }
+        List<List<Object>> argLists = generateCombinations(0, args);
+        Object[] instances = new Object[3];
+        try {
+            instances[0] = cons.newInstance(argLists.get(0).toArray());
+            instances[1] = cons.newInstance(argLists.get(0).toArray());
+            List<Object> otherInstances = new ArrayList<>();
+            for (List<Object> argList : argLists.subList(1, argLists.size())) {
+                otherInstances.add(cons.newInstance(argList.toArray()));
+            }
+            instances[2] = otherInstances.toArray();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        return instances;
+    }
+
+    private static List<List<Object>> generateCombinations(int index, List<List<Object>> args) {
+        List<List<Object>> result = new ArrayList<>();
+        if (index == args.size()) {
+            result.add(new ArrayList<>());
+        } else {
+            for (Object obj : args.get(index)) {
+                for (List<Object> list : generateCombinations(index + 1, args)) {
+                    list.add(0, obj);
+                    result.add(list);
+                }
+            }
+        }
+        return result;
+    }
+
+    public static Object OTHER_TYPE = new Object() {};
+
+    private final Object object;
+    private final Object same;
+    private final Object[] other;
+
+    public AutoEqualityTest(final Object object, final Object same, final Object[] other) throws NoSuchMethodException {
```

Comment:
Wow... This has lead me to also remove a huge amount of constructors from all `ParameterizedParse` subclasses :)